### ISSUE TITLE
Tab key width vs Tab character width

### ISSUE THREAD

**Issue by DarkArc** [zed-industries/zed#14080]
### Summary

I'm working in a code base where tabs are 8 spaces and should be rendered as such. However, the typical indent within a block scope is 2 spaces.

It would be nice if there was a separate setting for the width of a tab character and how many spaces are typed when the tab key is pressed.

**Comment by notpeter** [comment]
Is this something that would be handled by the [tab_size](https://zed.dev/docs/configuring-zed#tab-size) configuration option? You can set this globally and/or override on a per language basis.

**Comment by DarkArc** [comment]
@notpeter no, the tab_size doesn't affect the rendered width. Even with:

```
      "hard_tabs": true,
      "tab_size": 8,
```

Code that should look like this:
```
        int     hi;     /* Comment about hi
                           that continues. */
```

Looks like this:
```
  int     hi;     /* Comment about hi
         that continues. */
```

**Comment by ArtemiyGranat** [comment]
Hi!

Thank you for the excellent editor, I really enjoy using it, but the problem is that one of my
projects uses the GNU code style, where the tab takes up 8 spaces and the indentation is 2
spaces. AFAIK, Zed doesn't support GNU code style right now, so I can't use it as my main editor.

I have two questions:
1. What is the priority of this task?
2. Are there any workarounds to solve this problem temporarily?

**Comment by notpeter** [comment]
@ArtemiyGranat Am I correct, you want to hard_tabs (\t) with a tab_size of 8, but have the actual tab key insert two spaces and not a hard tab?

@DarkArc Here is my unsuccessful attempt to reproduce your rendering issue:

<details><summary>code</summary>

```sh 
mkdir -p /tmp/issue14080/.zed/
cd /tmp/issue14080
echo '{
  "languages": {
    "C": {
      "show_whitespaces": "all",
      "hard_tabs": true,
      "tab_size": 8
    },
}' > .zed/settings.json

echo 'int             var_a;  /* Comment about var_a
                           that continues. */

int		var_b;
			/* Comment about var_b. */

enum x : int {
  A,
  B,
  C
};

int foo() {
  int c = var_a;
  if (c == 10) {
    return var_b;
  }
  return c;
}' > a.c
zed . a.c
```

</details>

Which renders as:

<img width="590" alt="Image" src="https://github.com/user-attachments/assets/ac416d0e-c1cd-4e2f-b7d2-35c792b920a0" />

I cannot reproduce the rendering issue you are suggesting is occurring (tabs size being rendered as 2)? Can you provide explicit modification of my steps which triggers the behavior (or point out in the image where it looks wrong)?

Do you perhaps have .editorconfig in your project (or one of the parent folders like your home directory) which is setting `tab_size` = 2? If you run my repro steps above do you get the same visual? (notably it's under /tmp so there is unlikely to be a stray .editorconfig)

> It would be nice if there was a separate setting for the width of a tab character and how many spaces are typed when the tab key is pressed.
> In Emacs I currently use:
> <tab> = auto indent
> <C-tab> = insert 8 spaces
> <C-q C-i> = insert a tab character

So to be clear, you're looking for a setting where `hard_tabs` is true, but the tab key inserts spaces two spaces, ctrl-tab inserts 8 and `ctrl-q ctrl-i` lets you insert a hard tab:

I think this should work if you add it to your user keymap:
```json
[
  {
    "context": "Editor",
    "bindings": {
      "tab": [
        "workspace::SendKeystrokes",
        "space space"
      ],
      "ctrl-tab": [
        "workspace::SendKeystrokes",
        "space space space space space space space space"
      ],
      "ctrl-q ctrl-i": "editor::Tab"
    }
  }
]
```

I'm still trying to flesh out what the explicit request is here, it sounds like you're looking for settings which make it possible to have distinct keys for hard_tab, 2space_tab and auto-indent. In particular the behavior of auto-indent in a mixed hard_tabs / two space world seems a bit hard to define. Is this something where I can observe similar behavior in existing editors (in addition your emacs example)?

**Comment by ArtemiyGranat** [comment]
@notpeter Yes, you are right. You can see sample `.editorconfig` file in https://gcc.gnu.org/wiki/FormattingCodeForGCC:
```
[*.cc]
indent_style = tab
indent_size = 2
tab_width = 8
end_of_line = lf
charset = utf-8
trim_trailing_whitespace = true
insert_final_newline = true
```

Zed will not render projects with such `.editoconfig` correctly.

**Comment by darrinsmart** [comment]
Adding my 2c.. Where I work we have tab width of 8 and an indent of 4.

The tab width (8) is used when rendering text. A tab character moves to the next 8-column boundary.

The indent width (4) is used when indenting/outdenting text or pressing the tab key.

This means that the first indent, to column 5 must be done with spaces. The second indent, to column 9 can made with either spaces or a single tab, depending on user preference.

In Xcode this is done with the settings:

- Tab Width
- Indent Width
- Prefer Indent Using

In terms of editorconfig, the [relevant settings](https://spec.editorconfig.org/#supported-pairs) are:

- indent_style
- indent_size
- tab_width

Zed has `hard_tabs` (editorconfig: indent_style) and `tab_size` (editorconfig: tab_width). I think it's just missing an `indent_width` (editorconfig: indent_size) setting.

Indent guides would be drawn to `indent_width` intervals, not `tab_width`.

**Comment by notpeter** [comment]
Wonderful.  I had a hunch there was something else going on here that I was missing. Thanks for following up.

Closing in favor of this issue because it has a clearer narrative:
- https://github.com/zed-industries/zed/issues/24108

Please 👍 and subscribe to that issue.
Thanks for reporting.
